File: Top_Interview_Questions/addTwoNumbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    if l1 == nil && l2 == nil {
        return nil
    }
    head := &ListNode{}
    l1Val := 0
    l2Val := 0
    if l1 != nil {
        l1Val = l1.Val
    }
    if l2 != nil {
        l2Val = l2.Val
    }

    sum := l1Val + l2Val
    remainder := 0
    if sum >= 10 {
        sum -= 10
        remainder = 1
    }

    head.Val = sum

    if l1 == nil {
        if l2.Next != nil {
            l2.Next.Val += remainder
            head.Next = addTwoNumbers(nil, l2.Next)
        } else if remainder == 1 {
            head.Next = addTwoNumbers(&ListNode{1, nil}, nil)
        }
    } else if l2 == nil {
        if l1.Next != nil {
            l1.Next.Val += remainder
            head.Next = addTwoNumbers(l1.Next, nil)
        } else if remainder == 1 {
            head.Next = addTwoNumbers(&ListNode{1, nil}, nil)
        }
    } else {
        if remainder == 1 {
            if l1.Next != nil {
                l1.Next.Val++
                head.Next = addTwoNumbers(l1.Next, l2.Next)
            } else if l2.Next != nil {
                l2.Next.Val++
                head.Next = addTwoNumbers(l1.Next, l2.Next)
            } else {
                head.Next = addTwoNumbers(&ListNode{1, nil}, nil)
            }
        } else {
            head.Next = addTwoNumbers(l1.Next, l2.Next)
        }
    }
    return head
}

File: misc/add_two_numbers.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        _sum = 0
        carry = 0
        cur_1 = l1
        cur_2 = l2
        new_head = ListNode(0, None)
        cur = new_head
        
        while cur_1 or cur_2:
            if cur_1 and cur_2:
                _sum = (cur_1.val + cur_2.val + carry)
            elif cur_1:
                _sum = (cur_1.val + carry)
            else:
                _sum = (cur_2.val + carry)
            new_val = _sum % 10
            carry = _sum // 10
            cur.val = new_val
            if (cur_1 and cur_1.next) or (cur_2 and cur_2.next):
                cur.next = ListNode(0, None)
            elif carry:
                cur.next = ListNode(carry, None)
                break
            cur = cur.next
            cur_1 = cur_1.next if cur_1 else cur_1
                
            cur_2 = cur_2.next if cur_2 else cur_2
            
            
        return new_head